Source/common/SantaCache.h

```diff
@@ -146,6 +146,39 @@ class SantaCache {
     return set(key, zero_, update_block, {}, false);
   }
 
+  /**
+    Iterate all key and value pairs in the cache.
+
+    @param foreach_block Called for all key and value pairs.
+  */
+  void foreach (std::function<void(KeyT &, ValueT &)> foreach_block) {
+    assert(foreach_block != nullptr);
+
+    // Prevent the cache from being cleared while iterating
+    lock(&clear_bucket_);
```

Comment:
I'm not sure this is strictly necessary, though admittedly it probably doesn't hurt much.

The clear function locks every bucket before it clears and as you're locking all the buckets too, either you got a bucket lock and a clear will wait to grab it, or vice-versa. The clear bucket is to prevent back-to-back clears when the cache overflows.

Reply:
Yeah, this is true. Will remove.

If multiple threads call both functions simultaneously - whichever method is first to lock the first bucket, that method will then complete it's entire task before the other method can start. This works since, as you mention, both methods lock all buckets and, importantly, both start from index 0 and lock sequentially. I wanted to ensure that a cache couldn't be "partially cleared", but that can't happen here.